### ISSUE TITLE
Question: Completion in nvim's terminal window

### ISSUE THREAD

**Issue by jamilraichouni** [hrsh7th/nvim-cmp#1219]
Hi!

First: Thank you very much for the different completion/ source plugins @hrsh7th! The stuff is really helpful!

My question:
Is there any chance/ trick how I could achieve that `nvim-cmp` triggers in the internal terminal window?

Use cases:

* Shell script snippets in zsh shell
* Python snippets in IPython shell or Python shell
* In general stuff provided by the [`cmp-buffer`](https://github.com/hrsh7th/cmp-buffer)

That would be really awesome!

many thanks and all best,
Jamil

**Comment by Shougo** [comment]
I know it is useful.
But it is very hard.  Because neovim does not support it...

**Comment by jamilraichouni** [comment]
Thanks for the quick response. What exactly is not supported? Is it popup menu's in the terminal buffer? Or what else is missing?
Maybe it makes sense to raise an according feature request directly addressed at the Neovim contributors!?

**Comment by Shougo** [comment]
well, I had to implement the feature in pum.vim.
The terminal mode doesn't support get current cursor and it doesn't support the text insertion(need feedkeys).

**Comment by Shougo** [comment]
I will test it later...

**Comment by Shougo** [comment]
I have tested it and I found the critical problem.

In terminal mode, it does not have `CmdChanged`/`InsertChanged` autocmd.

**Comment by Shougo** [comment]
I have created the PR for it.

https://github.com/neovim/neovim/pull/20567

**Comment by yyy33** [comment]
r

> I have tested it and I found the critical problem.
> 
> In terminal mode, it does not have `CmdChanged`/`InsertChanged` autocmd.

For zsh, I'm not sure if this can be done with zle + $NVIM(unix socket) + user autocmd

**Comment by Shougo** [comment]
Well, it may be work.  But this is just hack.  I don't want to hack it.

**Comment by yyy33** [comment]
> Well, it may be work. But this is just hack. I don't want to hack it.

Yes, it would be best if it could be implemented in neovim, it looks like you are familiar with both neovim and nvim-cmp projects, I am a newbie, mind sharing how you read open source projects quickly?

**Comment by Shougo** [comment]
Hm... It is hard problem.
It needs very long experience.
I have worked for Vim in 15 years.

**Comment by Shougo** [comment]
The patch is already merged!

**Comment by jamilraichouni** [comment]
Great, so I just installed the latest nightly build for nvim. How can I get nvim-cmp to trigger and show any snippets?
A small example would be really cool.

Thank you very much @Shougo

**Comment by yyy33** [comment]
> Great, so I just installed the latest nightly build for nvim. How can I get nvim-cmp to trigger and show any snippets? A small example would be really cool.
> 
> Thank you very much @Shougo

I don't think nvim-cmp supports it yet, so you should manually add the trigger event TextChangedT to trigger the completion of nvim-cmp,   [https://github.com/hrsh7th/nvim-cmp/issues/519#issuecomment-968909772](url)

**Comment by yyy33** [comment]
> Great, so I just installed the latest nightly build for nvim. How can I get nvim-cmp to trigger and show any snippets? A small example would be really cool.
> 
> Thank you very much @Shougo

You may also need to install a source, such as zsh:
https://github.com/tamago324/cmp-zsh

**Comment by Shougo** [comment]
> I don't think nvim-cmp supports it yet, so you should manually add the trigger event TextChangedT to trigger the completion of nvim-cmp, https://github.com/hrsh7th/nvim-cmp/issues/519#issuecomment-968909772

Yes.  It needs `nvim-cmp`'s support.
And the terminal completion is not perfect.

**Comment by jamilraichouni** [comment]
Hi!

I tried it out and could not get a snippet popup to appear in the terminal.

So when I configure `nvim-cmp` correctly it should work? Or do we need to wait for some feature addition here in this repo?

What I tried:

I added the zsh source and I also put the snippet from https://github.com/hrsh7th/nvim-cmp/issues/519#issuecomment-969674104 into my init.vim. At the start of the lua function I added a print of the current time incl. seconds and always see an update of the time when I type something in the terminal. 

```vimL
autocmd TextChangedT * call s:on_complete_check()

function! s:on_complete_check() abort
lua <<EOF
 print(os.date("%X"))
 local line = vim.api.nvim_get_current_line()
 local cursor = vim.api.nvim_win_get_cursor(0)[2]

 local current = string.sub(line, cursor, cursor + 1)
 if current == "." or current == "," or current == " " then
    require('cmp').close()
 end

 local before_line = string.sub(line, 1, cursor + 1)
 local after_line = string.sub(line, cursor + 1, -1)
 if not string.match(before_line, '^%s+$') then
    if after_line == "" or string.match(before_line, " $") or string.match(before_line, "%.$") then
        require('cmp').complete()
    end
 end
EOF
endfunction
```

I get no snippet list to appear.
What I hope that it will be possible at some day (like a Christmas present) is that I can see zsh snippets in the terminal and Python snippets if I like to (running IPython in the terminal).

In fact I can put some wrapping code around the IPython command and do the switch in the background.

Hence, if I can get any zsh snippet to be shown, that would already be a big win.

**Comment by yyy33** [comment]
> 

Sorry, I don't know what the problem is, maybe we should wait for @Shougo's pl to solve the problem

**Comment by Shougo** [comment]
If you need the feature, `nvim-cmp` should support `TextChangedT` autocmd.
Unfortunately, it is not may work.

**Comment by Shougo** [comment]
@jamilraichouni I have tested your snippet.
`TextChangedT` works, but `require('cmp').complete()` does not work in terminal mode.

**Comment by Shougo** [comment]
```diff
diff --git a/lua/cmp/core.lua b/lua/cmp/core.lua
index 6ec11a0..2cde661 100644
--- a/lua/cmp/core.lua
+++ b/lua/cmp/core.lua
@@ -371,7 +371,7 @@ core.confirm = function(self, e, option, callback)
   feedkeys.call('', 'n', function()
     -- Restore the line at the time of request.
     local ctx = context.new()
-    if api.is_cmdline_mode() then
+    if api.is_cmdline_mode() or api.is_terminal_mode() then
       local keys = {}
       table.insert(keys, keymap.backspace(ctx.cursor_before_line:sub(e:get_offset())))
       table.insert(keys, string.sub(e.context.cursor_before_line, e:get_offset()))
diff --git a/lua/cmp/utils/api.lua b/lua/cmp/utils/api.lua
index 8153c9e..89f3075 100644
--- a/lua/cmp/utils/api.lua
+++ b/lua/cmp/utils/api.lua
@@ -13,6 +13,8 @@ api.get_mode = function()
     return 's' -- select
   elseif mode == 'c' and vim.fn.getcmdtype() ~= '=' then
     return 'c' -- cmdline
+  elseif mode == 't' then
+    return 't' -- terminal
   end
 end
 
@@ -32,9 +34,13 @@ api.is_visual_mode = function()
   return api.get_mode() == 'x'
 end
 
+api.is_terminal_mode = function()
+  return api.get_mode() == 't'
+end
+
 api.is_suitable_mode = function()
   local mode = api.get_mode()
-  return mode == 'i' or mode == 'c'
+  return mode == 'i' or mode == 'c' or mode == 't'
 end
 
 api.get_current_line = function()

```

Here is the minimal patch to support terminal mode.

**Comment by yyy33** [comment]
> ```diff
> diff --git a/lua/cmp/core.lua b/lua/cmp/core.lua
> index 6ec11a0..2cde661 100644
> --- a/lua/cmp/core.lua
> +++ b/lua/cmp/core.lua
> @@ -371,7 +371,7 @@ core.confirm = function(self, e, option, callback)
>    feedkeys.call('', 'n', function()
>      -- Restore the line at the time of request.
>      local ctx = context.new()
> -    if api.is_cmdline_mode() then
> +    if api.is_cmdline_mode() or api.is_terminal_mode() then
>        local keys = {}
>        table.insert(keys, keymap.backspace(ctx.cursor_before_line:sub(e:get_offset())))
>        table.insert(keys, string.sub(e.context.cursor_before_line, e:get_offset()))
> diff --git a/lua/cmp/utils/api.lua b/lua/cmp/utils/api.lua
> index 8153c9e..89f3075 100644
> --- a/lua/cmp/utils/api.lua
> +++ b/lua/cmp/utils/api.lua
> @@ -13,6 +13,8 @@ api.get_mode = function()
>      return 's' -- select
>    elseif mode == 'c' and vim.fn.getcmdtype() ~= '=' then
>      return 'c' -- cmdline
> +  elseif mode == 't' then
> +    return 't' -- terminal
>    end
>  end
>  
> @@ -32,9 +34,13 @@ api.is_visual_mode = function()
>    return api.get_mode() == 'x'
>  end
>  
> +api.is_terminal_mode = function()
> +  return api.get_mode() == 't'
> +end
> +
>  api.is_suitable_mode = function()
>    local mode = api.get_mode()
> -  return mode == 'i' or mode == 'c'
> +  return mode == 'i' or mode == 'c' or mode == 't'
>  end
>  
>  api.get_current_line = function()
> ```
> 
> Here is the minimal patch to support terminal mode.

@hrsh7th Can we merge this patch?
